// steam-searcher/index.js
var request = require('request')

function BuildAppIDArray(htmlSrc) {
    const findStr = 'data-ds-appid="(\d+)"';
    var arrayAppIDString = htmlSrc.match(findStr);
    
    if (arrayAppIDString && arrayAppIDString.isArray()) {
        let outArray = [];
        arrayAppIDString.forEach(elem => outArray.push(parseInt(elem)));
        return outArray;
    }

    return null;
};

function GetAppIDDetails(arrayIDs, callback) {
    let aggData = []
    arrayIDs.forEach(function(element) {
        var steamID = element.id,
        reqUrlGame = `http://store.steampowered.com/api/appdetails?appids=${steamID}`
        console.log('Processing ' + element.name + '...')
        request.get({
                url: reqUrlGame,
                timeout: timeout
            }, function (err, res, body) {
                if (err) return callback(err);
                if (res.statusCode !== 200) return callback(new Error('request failed (' + res.statusCode + ')'));
                if (!body) return callback(new Error('failed to get body content'));

                var gameData = JSON.parse(body)
                console.log('Pushing ' + gameData[steamID].data.name);
                aggData.concat(aggData.push(gameData[steamID].data))
                
                if (aggData.length === count) {
                    console.log('Operation terminated with SUCCESS.')
                    return callback(null, aggData);
                }
            }
        )
    })
};

function FireGeneralRequestFromText(timeout, reqUrl, callback) {
    request.get({
        url: reqUrl,
        timeout: timeout
    }, function (err, res, body) {

        if (err) return callback(err);
        if (res.statusCode !== 200) return callback(new Error('request failed (' + res.statusCode + ')'));
        if (!body) return callback(new Error('failed to get body content'));

        let appIDArray = BuildAppIDArray(body);
        if (appIDArray) {
            console.log('found text for ' + appIDArray.length + ' games on that page!')
            console.log("Fetching details...")
            GetAppIDDetails(appIDArray, callback);
        }

        console.log("No appID was found on the html response to getting that page: " + reqUrl);
    })
};

function FireGeneralRequestFromJSON(timeout, reqUrl, callback) {
    request.get({
        url: reqUrl,
        timeout: timeout
    }, function (err, res, body) {

        if (err) return callback(err);
        if (res.statusCode !== 200) return callback(new Error('request failed (' + res.statusCode + ')'));
        if (!body) return callback(new Error('failed to get body content'));

        // Parse body
        var res1 = JSON.parse(body)
        // for each item found, throw up a request to get them all
        count = res1.items.length;
        console.log('Parsing JSON for ' + count + ' games found...')
        GetAppIDDetails(res1.items)
    })
};

// Init the module
function findFixedUrl(callback) {
    if (typeof callback !== 'function')
        callback = function callback(err, result) {
            return err || result;
        };

        // to get first 100 top selling games, ask for 4 pages of 25 results
        const numberPagesMax = 4;
        const timeout = 1000;
        const reqUrl = `https://store.steampowered.com/tags/en/Turn-Based%20Strategy/#p=0&tab=TopSellers`;  // page 1
        //let reqUrl = `https://store.steampowered.com/search/?sort_by=Released_DESC&tags=9&category1=998`;
        // TODO: add requets for more pages of results with the following format
        //reqUrl = `https://store.steampowered.com/tags/en/Strategy#p=1&tab=TopSellers`;    // page 2

    return FireGeneralRequestFromText(timeout, reqUrl, callback);
};

function findJSON(game, callback) {
    if (typeof callback !== 'function')
        callback = function callback(err, result) {
            return err || result;
        };

    if (!game || typeof game !== 'object')
        return callback('invalid game');

    if (!game.search)
        return callback('missing search input');

    var timeout = 1000,
        search = game.search,
        reqUrl = `http://store.steampowered.com/api/storesearch/?term={${search}}&l=english&cc=US`;

    return FireGeneralRequestFromJSON(timeout, reqUrl, callback);
};

function findGenre(genre, callback) {
    var timeout = 1000,
    genre = genre.search,
    reqUrl = `https://store.steampowered.com/tags/en/${genre}#p=0&tab=ComingSoon`;

    return FireGeneralRequestFromText(timeout, reqUrl, callback);
};

module.exports.find = findFixedUrl;
module.exports.findGame = findJSON;
module.exports.findGenre = findGenre;
